hack/verify-kube-api-linter.sh

```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env bash
+
+# Copyright 2021 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script checks the coding style for the Go language files using
+# golangci-lint. Which checks are enabled depends on command line flags. The
+# default is a minimal set of checks that all existing code passes without
+# issues.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+source "${KUBE_ROOT}/hack/lib/init.sh"
+source "${KUBE_ROOT}/hack/lib/util.sh"
+
+kube::golang::setup_env
+export GOBIN="${KUBE_OUTPUT_BIN}"
+
+echo "installing golangci-lint into ${GOBIN}"
+GOTOOLCHAIN="$(kube::golang::hack_tools_gotoolchain)" go -C "${KUBE_ROOT}/hack/tools" install github.com/golangci/golangci-lint/cmd/golangci-lint
+
+golangci_config="${KUBE_ROOT}/hack/kube-api-linter.yaml"
+kube_api_linter="${GOBIN}/kube-api-linter"
+
+if [ "${golangci_config}" ]; then
+  # The relative path to _output/local/bin only works if that actually is the
+  # GOBIN. If not, then we have to make a temporary copy of the config and
+  # replace the path with an absolute one. This could be done also
+  # unconditionally, but the invocation that is printed below is nicer if we
+  # don't to do it when not required.
+  if grep -q 'path: ../_output/local/bin/' "${golangci_config}" &&
+     [ "${GOBIN}" != "${KUBE_ROOT}/_output/local/bin" ]; then
+    kube::util::ensure-temp-dir
+    patched_golangci_config="${KUBE_TEMP}/$(basename "${golangci_config}")"
+    sed -e "s;path: ../_output/local/bin/;path: ${GOBIN}/;" "${golangci_config}" >"${patched_golangci_config}"
+    golangci_config="${patched_golangci_config}"
+  fi
+fi
+
+cd "${KUBE_ROOT}/hack"
+
+echo "installing kube-api-linter into ${GOBIN}"
+GOTOOLCHAIN="$(kube::golang::hack_tools_gotoolchain)" ${GOBIN}/golangci-lint custom
```

Comment:
I experimented with `golangci-lint custom` in https://github.com/kubernetes/kubernetes/pull/131507.

It has some drawbacks:
- golangci-lint needs to be built twice.
- Under the hood, `golangci-lint custom` checks out the Git repo again.
- Versioning in `.custom-gcl.yaml` is special.

We have sorted out how to do plugins as Go modules. Would it perhaps make more sense to keep using that also for kube-api-linter?

Reply:
https://github.com/kubernetes/kubernetes/pull/131561#pullrequestreview-2810429586:
> I think we still have some issues there with dockerized vs not.

I believe https://github.com/kubernetes/kubernetes/blame/03a3c0c89161935bc2338f5754ebb1104f779af1/hack/verify-golangci-lint.sh#L148-L162 took care of that.

Reply:
Thanks for the hints, I've got this switched over now to the same pattern as logcheck, should be simpler!